docs/RFCS/leader_evaluated_raft.md

```diff
@@ -0,0 +1,425 @@
+  // stable, i.e. non-commutative (overlapping writes) can't rearrange
+  // themselves freely). [i.e. can use a map under the hood]
+  // The returned Writes are read-only and must not be used after their Batch
+  // is finalized. [i.e. can reclaim the memory at that point, or we add an
+  // explicit method to finalize the writes if that is more convenient].
+  GetWrites() (Writes, error)
+}
+```
+
+With this strategy, the MVCC changes are fairly locally scoped and not too
+invasive.
+
+Naively, this can be achieved by wrapping a "regular" batch with an appropriate
+in-memory map which is populated on writes.
+
+However, using [RocksDB's `WriteBatch`
+format](https://github.com/facebook/rocksdb/blob/f38540b12ada4fe06598c42d4c084f9d920289ff/db/write_batch.cc#L10)
+seems opportune:
+
+```
+// WriteBatch::rep_ :=
+//    sequence: fixed64
+//    count: fixed32
+//    data: record[count]
+// record :=
+//    kTypeValue varstring varstring
+//    kTypeDeletion varstring
+//    kTypeSingleDeletion varstring
+//    kTypeMerge varstring varstring
+//    kTypeColumnFamilyValue varint32 varstring varstring
+//    kTypeColumnFamilyDeletion varint32 varstring varstring
+//    kTypeColumnFamilySingleDeletion varint32 varstring varstring
+//    kTypeColumnFamilyMerge varint32 varstring varstring
+// varstring :=
+//    len: varint32
+//    data: uint8[len]
+```
+
+* The format is straightforward to implement; the writes need to be serialized
+  anyway for the Raft proposal, and this appears to be an efficient format to
+  use, even when the underlying storage engine isn't based on RocksDB.
+* We currently use RocksDB and this is already the internal format used by a
+  RocksDB batch. We can completely avoid an additional copy and serialization
+  step in this case (i.e. construct the batch, borrow out the contained
+  representation to Raft until the command commits, and apply the batch).
+  Followers can create a `WriteBatch` directly from this representation.
+  Serialization is a key point in performance[1].
+* Again for RocksDB, we might also be able to piggy-back on existing
+  implementation to operate on the superposition of multiple `WriteBatch`es
+  (which would be needed to remove write-write blocking).
+
+The format should be relatively stable (likely only additions), but there might
+be future migrations coming up as RocksDB changes their underlying assumptions.
+We should consider checking in with them about this idea.
+
+Using `WriteBatch` also means that the writes are completely opaque to the
+stack (at least unless unserialized early, and loaded, and then accessed
+through MVCC operations), so the followers essentially can't act based on
+them. That should not be a problem (but could if we want the side effects
+to use information from the write set in a performant way).
+
+## Raft command application
+
+This will be a new, straightforward facility which applies a `RaftCmd` against
```

Comment:
I propose that we send the serialized bytes from the `WriteBatch` as a `bytes` parameter in `RaftCmd` and then pass that directly to a new cgo shim for our engine which would deserialize as necessary down in C++ land. In C++, we'd fork the current RocksDB `WriteBatch` protobuf and use that in our C++ `Batch` implementation. As long as we're using RocksDB, it'll be a trivial assignment of bytes. Other engines would require a translation. But by forking the RocksDB `WriteBatch`, we'll stabilize our own version as well as keep all of these details opaque to the Go implementation.